# Visualization

The first step is to visualize the data table contained in "P_DEMO.xpt". The first five observations are shown below.

In [2]:
import pandas as pd

df_p_demo = pd.read_sas("P_DEMO.xpt")

df_p_demo.head()

,SEQN,SDDSRVYR,RIDSTATR,RIAGENDR,RIDAGEYR,RIDAGEMN,RIDRETH1,RIDRETH3,RIDEXMON,DMDBORN4,...,FIAINTRP,MIALANG,MIAPROXY,MIAINTRP,AIALANGA,WTINTPRP,WTMECPRP,SDMVPSU,SDMVSTRA,INDFMPIR
0,109263.0,66.0,2.0,1.0,2.0,NaN,5.0,6.0,2.0,1.0,...,2.0,NaN,NaN,NaN,NaN,7891.762435,8.951816e+03,3.0,156.0,4.66
1,109264.0,66.0,2.0,2.0,13.0,NaN,1.0,1.0,2.0,1.0,...,2.0,1.0,2.0,2.0,1.0,11689.747264,1.227116e+04,1.0,155.0,0.83
2,109265.0,66.0,2.0,1.0,2.0,NaN,3.0,3.0,2.0,1.0,...,2.0,NaN,NaN,NaN,NaN,16273.825939,1.665876e+04,1.0,157.0,3.06
3,109266.0,66.0,2.0,2.0,29.0,NaN,5.0,6.0,2.0,2.0,...,2.0,1.0,2.0,2.0,1.0,7825.646112,8.154968e+03,2.0,168.0,5.00
4,109267.0,66.0,1.0,2.0,21.0,NaN,2.0,2.0,NaN,2.0,...,2.0,NaN,NaN,NaN,NaN,26379.991724,5.397605e-79,1.0,156.0,5.00


Each of its attributes is displayed, along with the number of rows and columns that comprise the dataset:

In [3]:
df_p_demo.columns

Index(['SEQN', 'SDDSRVYR', 'RIDSTATR', 'RIAGENDR', 'RIDAGEYR', 'RIDAGEMN',
       'RIDRETH1', 'RIDRETH3', 'RIDEXMON', 'DMDBORN4', 'DMDYRUSZ', 'DMDEDUC2',
       'DMDMARTZ', 'RIDEXPRG', 'SIALANG', 'SIAPROXY', 'SIAINTRP', 'FIALANG',
       'FIAPROXY', 'FIAINTRP', 'MIALANG', 'MIAPROXY', 'MIAINTRP', 'AIALANGA',
       'WTINTPRP', 'WTMECPRP', 'SDMVPSU', 'SDMVSTRA', 'INDFMPIR'],
      dtype='str')

In [4]:
df_p_demo.shape

(15560, 29)

The nomenclature used in the P_DEMO.xpt file is detailed below:

(This step is not automated and therefore applies only to the P_BIOPRO.xpt file)
<ol>
    <li><b>SEQN</b>: Respondent Sequence Number</li>
    <li><b>SDDSRVYR</b>: Data release cycle</li>
    <li><b>RIDSTATR</b>: Interview/Examination status</li>
    <li><b>RIAGENDR</b>: Gender</li>
    <li><b>RIDAGEYR</b>: Age in years at screening</li>
    <li><b>RIDAGEMN</b>: Age in months at screening - 0 to 24 mos</li>
    <li><b>RIDRETH1</b>: Race/Hispanic origin</li>
    <li><b>RIDRETH3</b>: Race/Hispanic origin w/ NH Asian</li>
    <li><b>RIDEXMON</b>: Six-month time period</li>
    <li><b>DMDBORN4</b>: Country of birth</li>
    <li><b>DMDYRUSZ</b>: Length of time in US</li>
    <li><b>DMDEDUC2</b>: Education level - Adults 20+</li>
    <li><b>DMDMARTZ</b>: Marital status</li>
    <li><b>RIDEXPRG</b>: Pregnancy status at exam</li>
    <li><b>SIALANG</b>: Language of SP Interview</li>
    <li><b>SIAPROXY</b>: Proxy used in SP Interview?</li>
    <li><b>SIAINTRP</b>: Interpreter used in SP Interview?</li>
    <li><b>FIALANG</b>: Language of Family Interview</li>
    <li><b>FIAPROXY</b>: Proxy used in Family Interview?</li>
    <li><b>FIAINTRP</b>: Interpreter used in Family Interview?</li>
    <li><b>MIALANG</b>: Language of MEC Interview</li>
    <li><b>MIAPROXY</b>: Proxy used in MEC Interview?</li>
    <li><b>MIAINTRP</b>: Interpreter used in MEC Interview?</li>
    <li><b>AIALANGA</b>: Language of ACASI Interview</li>
    <li><b>WTINTPRP</b>: Full sample interview weight</li>
    <li><b>WTMECPRP</b>: Full sample MEC exam weight</li>
    <li><b>SDMVPSU</b>: Masked variance pseudo-PSU</li>
    <li><b>SDMVSTRA</b>: Masked variance pseudo-stratum</li>
    <li><b>INDFMPIR</b>: Ratio of family income to poverty</li>
</ol>

# Data Cleaning

In the "P_DEMO.xpt" dataset, some columns contain repeated text-based comments, making them unsuitable for analysis since numerical data is required. Therefore, the dataset will be modified by removing these columns to improve data processing and analysis.

Additionally, the removal of rows containing only null values, except for the SEQN identifier, represents a data cleaning step that prepares the file for further processing with the other classifiers.


In [5]:
import pandas as pd

# 1. Load the .xpt file
input_name = 'P_DEMO.xpt'
df = pd.read_sas(input_name)

# --- STEP A: REMOVE SPECIFIC COLUMNS ---
columns_to_remove = [
    'SDDSRVYR', 'RIDSTATR', 'RIDAGEMN',
       'RIDRETH1', 'RIDRETH3', 'RIDEXMON', 'DMDBORN4', 'DMDYRUSZ', 'DMDEDUC2',
       'DMDMARTZ', 'RIDEXPRG', 'SIALANG', 'SIAPROXY', 'SIAINTRP', 'FIALANG',
       'FIAPROXY', 'FIAINTRP', 'MIALANG', 'MIAPROXY', 'MIAINTRP', 'AIALANGA',
       'WTINTPRP', 'WTMECPRP', 'SDMVPSU', 'SDMVSTRA', 'INDFMPIR'
]
df_limpio = df.drop(columns=columns_to_remove, errors='ignore')

# --- STEP A.1: REMOVE COLUMNS IF THE ENTIRE COLUMN IS NULL ---
# axis=1 indicates columns, how='all' indicates that ALL values must be NaN
df_limpio = df_limpio.dropna(axis=1, how='all')

# --- STEP B: REMOVE EMPTY ROWS ---
# Define which columns will be used to evaluate whether the row is "empty"
# (Using the current columns of df_limpio)
data_columns = [col for col in df_limpio.columns if col != 'SEQN']

# how="all" means that a row is removed only if ALL data columns are NaN
df_limpio = df_limpio.dropna(subset=data_columns, how='all')

# --- STEP C: SAVE RESULTS ---
output_csv_name = 'P_DEMO_limpio.csv'
df_limpio.to_csv(output_csv_name, index=False, encoding='utf-8-sig')

# --- CHANGE REPORT ---
print("-" * 30)
print("FINAL CLEANING REPORT")
print("-" * 30)
print(f"Original rows: {len(df)}")
print(f"Resulting rows: {len(df_limpio)}")
print(f"Original columns: {len(df.columns)}")
print(f"Columns removed by list: {len(columns_to_remove)}")
print(f"Final columns: {len(df_limpio.columns)}")
print(f"File saved as: {output_csv_name}")

------------------------------
FINAL CLEANING REPORT
------------------------------
Original rows: 15560
Resulting rows: 15560
Original columns: 29
Columns removed by list: 26
Final columns: 3
File saved as: P_DEMO_limpio.csv


Initially, the dataset contained 15,560 rows and 29 columns. After the data cleaning process, the dataset was reduced to 15,560 rows and 3 columns, containing effective data suitable for further analysis without any issues. The process generates a new dataset named "P_DEMO_limpio.csv". An important aspect to note is that the file extension was changed to a more accessible and user-friendly format.

In [6]:
import pandas as pd

df_p_demo_limpio = pd.read_csv("P_DEMO_limpio.csv")

df_p_demo_limpio.head()

,SEQN,RIAGENDR,RIDAGEYR
0,109263.0,1.0,2.0
1,109264.0,2.0,13.0
2,109265.0,1.0,2.0
3,109266.0,2.0,29.0
4,109267.0,2.0,21.0


Additionally, when comparing the columns of the original file with those of the new file, a considerable reduction in attributes can be observed, resulting in 25 attributes available for further analysis.

The following section presents a comparison of the attributes and dimensions of the original and cleaned datasets from the "P_DEMO.xpt" file:

In [7]:
df_p_demo.columns

Index(['SEQN', 'SDDSRVYR', 'RIDSTATR', 'RIAGENDR', 'RIDAGEYR', 'RIDAGEMN',
       'RIDRETH1', 'RIDRETH3', 'RIDEXMON', 'DMDBORN4', 'DMDYRUSZ', 'DMDEDUC2',
       'DMDMARTZ', 'RIDEXPRG', 'SIALANG', 'SIAPROXY', 'SIAINTRP', 'FIALANG',
       'FIAPROXY', 'FIAINTRP', 'MIALANG', 'MIAPROXY', 'MIAINTRP', 'AIALANGA',
       'WTINTPRP', 'WTMECPRP', 'SDMVPSU', 'SDMVSTRA', 'INDFMPIR'],
      dtype='str')

In [8]:
df_p_demo_limpio.columns

Index(['SEQN', 'RIAGENDR', 'RIDAGEYR'], dtype='str')

In [9]:
df_p_demo.shape

(15560, 29)

In [10]:
df_p_demo_limpio.shape

(15560, 3)

# Renaming to More Readable Names

Ahora, se cambian los nombres de los atributos a unos mas legibles para comodidad de estudio.

In [12]:
import numpy as np

# 1. Mapping dictionary
mapping_demo = {
    'SEQN': 'Participant_ID',
    'SDDSRVYR': 'Data release cycle',
    'RIDSTATR': 'Interview/Examination status',
    'RIAGENDR': 'Gender',
    'RIDAGEYR': 'Age',
    'RIDAGEMN': 'Age in months at screening - 0 to 24 mos',
    'RIDRETH1': 'Race/Hispanic origin',
    'RIDRETH3': 'Race/Hispanic origin w/ NH Asian',
    'RIDEXMON': 'Six-month time period',
    'DMDBORN4': 'Country of birth',
    'DMDYRUSZ': 'Length of time in US',
    'DMDEDUC2': 'Education level - Adults 20+',
    'DMDMARTZ': 'Marital status',
    'RIDEXPRG': 'Pregnancy status at exam',
    'SIALANG': 'Language of SP Interview',
    'SIAPROXY': 'Proxy used in SP Interview?',
    'SIAINTRP': 'Interpreter used in SP Interview?',
    'FIALANG': 'Language of Family Interview',
    'FIAPROXY': 'Proxy used in Family Interview?',
    'FIAINTRP': 'Interpreter used in Family Interview?',
    'MIALANG': 'Language of MEC Interview',
    'MIAPROXY': 'Proxy used in MEC Interview?',
    'MIAINTRP': 'Interpreter used in MEC Interview?',
    'AIALANGA': 'Language of ACASI Interview',
    'WTINTPRP': 'Full sample interview weight',
    'WTMECPRP': 'Full sample MEC exam weight',
    'SDMVPSU': 'Masked variance pseudo-PSU',
    'SDMVSTRA': 'Masked variance pseudo-stratum',
    'INDFMPIR': 'Ratio of family income to poverty'
}

# 2. Rename the columns
df_p_demo_final = df_p_demo_limpio.rename(columns=mapping_demo)

# --- AGE TREATMENT ---
# First: Convert individuals younger than 1 year to NaN
df_p_demo_final.loc[df_p_demo_final['Age'] < 1, 'Age'] = np.nan

# Second: Remove all rows where 'Age' is NaN
# (This removes original missing values and newly created ones)
df_p_demo_sin_nulos = df_p_demo_final.dropna(subset=['Age'])
# ---------------------------

# 3. Export to a new CSV file
demo_file_name = 'P_DEMO_nombres_claros.csv'
df_p_demo_sin_nulos.to_csv(demo_file_name, index=False, encoding='utf-8-sig')

# Quick verification
removed_rows = len(df_p_demo_final) - len(df_p_demo_sin_nulos)
print(f"✅ Process completed.")
print(f"Removed rows (Age < 1 or missing): {removed_rows}")
print(f"Total final records: {len(df_p_demo_sin_nulos)}")

✅ Process completed.
Removed rows (Age < 1 or missing): 574
Total final records: 14986


The generated dataset is displayed again for visualization and inspection.

In [13]:
import pandas as pd

df_p_demo_nombres_claros = pd.read_csv("P_DEMO_nombres_claros.csv")

df_p_demo_nombres_claros.head()

,Participant_ID,Gender,Age
0,109263.0,1.0,2.0
1,109264.0,2.0,13.0
2,109265.0,1.0,2.0
3,109266.0,2.0,29.0
4,109267.0,2.0,21.0


This concludes the data cleaning process for the P_DEMO.xpt file.

A cleaned and more readable dataset named "P_DEMO_nombres_claros.csv" has been obtained. This file will be used as input for further processing in a subsequent script.